### PR TITLE
[PORT] Bot Multi-Z Movement 

### PR DESCRIPTION
**Pull request by Bobbanz1**
## About The Pull Request
**Ports the Bot Multi-Z Movement from NSV13**

While this PR is no doubt useless for Beestation, given how they are using a single z-level station, it might be of use for downstreams who do use multi-z stations.

While this PR is capable of being used on any ship that has a bot elevator on it, if the AI is the one doing the summoning, however it can't do the Multi-Z Patrol thing just yet, as that requires all ships to have their navigation beacons reprogrammed from going in one big loop to instead going from destination to destination and then finally up a ladder, after which the next deck will also need a route that has a navigation beacon programmed to make the bot go downwards. 

(This can easily be implemented by copying over the location name of a beacon downstairs, as the next_patrol destination of some navbeacon upstairs)

Ports some of the code introduced by this PR: [tgstation/tgstation/pull/65906](https://github.com/tgstation/tgstation/pull/65906)
Full credit for the code used to find the closest ladder goes to cacogen.
Although the code has been modified

- [x] Implements the ability for bots to go between z-levels
- [x] Implement the ability for bots to patrol through z-levels
- [x] Implement the ability to summon bots from other z-levels
- [x] Implement the ability for the AI to call bots to other z-levels  
- [x] Implement this ability for Mulebots to deliver stuff between z-levels

Second of all!
Some bots, possibly all of them, are now capable of moving between Z-levels if one of the following conditions are met:

- The AI has slammed down a waypoint on another Z-level, like on the bottom deck or the upper deck.
- - If this happens then the bot will instantly head for the nearest bot elevator it can find from its current position, upon arriving to this bot elevator it will check the original waypoint and use the bot elevator to either go up or down, afterwards it resets the bot call command before promptly running it yet again, this time with the original waypoint as the final destination.
- - - This annoyingly produces a text message in the chat saying that the call command has been reset.
- A Navigation Beacon has a destination that's located on another Z-level, currently it's restricted to only check Z-level 2 and 3 which seems to always be the Z-levels the ships various decks are located on. 
- - Upon detecting the beacons location on another deck, the bot will try heading for the nearest bot elevator possible.
- - I've introduced a new structure which uses the same code as the ladders however this thing is specifically designed to be used by bots and will reject anyone else trying to use it to go up or down, this thing is called a bot elevator, the reason I introduced this thing is because...well it's easier for mappers to just replicate what they do with ladders for these things and then no longer have to worry about it.
- - - Upon reaching the bot elevator it will go down, or up it, and promptly forget the original destination and start heading for the nearest beacon it can find (could possibly also be trying to reach the next beacon from the original destination one)
- A PDA has Summoned this bot.

Third of all!
The Bot Elevators
They're currently using a temporary upside down version of the bin icon from the mining folder and act the same as ladders except for only being usable by bots, these things are created because *cough*_Moisturizer_*cough* wanted ladders for bots only. 

Anyways for any future mappers reading this, basically just do the same thing you do with ladders for these things.

Currently only the Atlas has a somewhat semi-working patrol network, although it sometimes gets stuck upstairs if the bot originally started its route on the navigation beacon outside the Head of Security's office. 

And now for the horrible, _horrible_ reveal, these bots are using KNPC Z-level travel to be capable of doing this stuff.

## Why It's Good For The Game
Bot uses bot elevators. Bot is now advanced boi.

## Changelog
:cl: 
add: Added the ability for bots to go up and down bot elevators.
add: Added the ability for bots to find a route to another Z-level if called to it.
add: Added the ability for bots to have a patrol route going downstairs or upstairs using bot elevators.
add: Added the ability for bots to be summoned using a PDA on another z-level.
add: Added the ability for mulebots to travel between Z-levels.
add: Added the ability to summon bots from other station Z-levels in the bot control computer program
/:cl:

### CODE REVIEW

File: code/modules/mob/living/simple_animal/bot/bot.dm [hunk 1]
```diff
@@ -75,6 +75,11 @@
 	var/blockcount = 0		//number of times retried a blocked path
 	var/awaiting_beacon	= 0	// count of pticks awaiting a beacon response
 
+	var/turf/last_waypoint
+	var/bot_z_mode //SETTINGS: 10 = AI CALLED. 20 = PATROLLING. 30 = SOMEONE CALLED.
```
Comment (itsmeow) [inline code/modules/mob/living/simple_animal/bot/bot.dm]: Add DEFINES please

File: code/modules/mob/living/simple_animal/bot/bot.dm [hunk 2]
```diff
@@ -685,27 +705,32 @@ Pass a positive integer as an argument to override a bot's default speed.
...
 /mob/living/simple_animal/bot/proc/get_next_patrol_target()
 	// search the beacon list for the next target in the list.
-	for(var/obj/machinery/navbeacon/NB in GLOB.navbeacons["[z]"])
-		if(NB.location == next_destination) //Does the Beacon location text match the destination?
-			destination = new_destination //We now know the name of where we want to go.
-			patrol_target = NB.loc //Get its location and set it as the target.
-			next_destination = NB.codes["next_patrol"] //Also get the name of the next beacon in line.
-			return TRUE
+	var/temporary_z = SSmapping.levels_by_trait(ZTRAIT_STATION)
+	for(var/ZYes in temporary_z)
```
Comment (itsmeow) [inline code/modules/mob/living/simple_animal/bot/bot.dm]: what is this awful var name, also if temporary_z is unused don't define the list var

File: code/game/objects/items/devices/PDA/cart.dm [hunk 3]
```diff
@@ -672,10 +672,16 @@ Code:
 		var/botcount = 0
 		for(var/B in GLOB.bots_list) //Git da botz
 			var/mob/living/simple_animal/bot/Bot = B
-			if(!Bot.on || Bot.get_virtual_z_level() != zlevel || Bot.remote_disabled || !(bot_access_flags & Bot.bot_type)) //Only non-emagged bots on the same Z-level are detected!
+			if(!Bot.on || Bot.remote_disabled || !(bot_access_flags & Bot.bot_type)) //Only non-emagged bots are detected!
 				continue //Also, the PDA must have access to the bot type.
-			menu += "<A href='byond://?src=[REF(src)];op=control;bot=[REF(Bot)]'><b>[Bot.name]</b> ([Bot.get_mode()])<BR>"
-			botcount++
+			if(Bot.get_virtual_z_level() in SSmapping.levels_by_trait(ZTRAIT_STATION))
```
Comment (itsmeow) [inline code/game/objects/items/devices/PDA/cart.dm:677]: This probably needs to be added to the Bot Control app for modular computers as well.

File: code/modules/mob/living/simple_animal/bot/bot.dm [hunk 4]
```diff
@@ -782,6 +823,10 @@ Pass a positive integer as an argument to override a bot's default speed.
 		return
 
 	if(loc == summon_target)		// Arrived to summon location.
+		if(last_summon != null)
+			if(z > last_summon.z || z < last_summon.z)
```
Comment (itsmeow) [inline code/modules/mob/living/simple_animal/bot/bot.dm]: z != last_summon.z?

File: code/modules/mob/living/simple_animal/bot/bot.dm [hunk 5]
```diff
@@ -1067,3 +1115,150 @@ Pass a positive integer as an argument to override a bot's default speed.
...
+			else if(z < last_summon.z)
+				E.travel(TRUE, src, FALSE, E.up, FALSE)
+				summon_target = last_summon
+				calc_summon_path()
+
+//BOT MULTI-Z MOVEMENT
+/mob/living/simple_animal/bot/proc/call_bot_z_move(caller, turf/ori_dest, message=TRUE)
+	//For giving the bot temporary all-access.
+	var/obj/item/card/id/all_access = new /obj/item/card/id
+	var/datum/job/captain/All = new/datum/job/captain
```
Comment (itsmeow) [inline code/modules/mob/living/simple_animal/bot/bot.dm]: weird capitalized var name

File: code/modules/mob/living/simple_animal/bot/medbot.dm [hunk 6]
```diff
@@ -108,6 +108,7 @@ GLOBAL_VAR(medibot_unique_id_gen)
 	. = ..()
 	var/datum/job/medical_doctor/J = new /datum/job/medical_doctor
 	access_card.access += J.get_access()
+	access_card.access += ACCESS_MAINT_TUNNELS
```
Comment (itsmeow) [inline code/modules/mob/living/simple_animal/bot/medbot.dm]: document in CL

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/simple_animal/bot/bot.dm (hunk 1 above)
Comment (Bobbanz1) [reply code/modules/mob/living/simple_animal/bot/bot.dm]: I'm unsure of what you want me to do here
Do you want me to just define them like 
#define bot_z_mode ?

File: code/modules/mob/living/simple_animal/bot/bot.dm (hunk 5 above)
Comment (Bobbanz1) [reply code/modules/mob/living/simple_animal/bot/bot.dm]: var/datum/job/captain/All is a direct copy from another section of this code, it's something that already exists in the bot code even before I started modifying it.

File: code/modules/mob/living/simple_animal/bot/bot.dm (hunk 1 above)
Comment (itsmeow) [reply code/modules/mob/living/simple_animal/bot/bot.dm]: #DEFINE BOT_Z_MODE_AI_CALLED 10
etc

Comment (itsmeow) [reply code/modules/mob/living/simple_animal/bot/bot.dm]: Should go in `code/_DEFINES/` in one of the fitting files, not sure which one. If needed make a new one.

**Review by itsmeow — APPROVED** [review]

**Review by PowerfulBacon — APPROVED** [review]

**Pull request merged**